File: code/915.分割数组.java
/*
 * @lc app=leetcode.cn id=915 lang=java
 *
 * [915] 分割数组
 */

// @lc code=start
class Solution {
    public int partitionDisjoint(int[] nums) {
        int[] ns = new int[nums.length];
        ns[0] = nums[0];
        int[] ms = new int[nums.length];
        ms[nums.length - 1] = nums[nums.length - 1];
        for (int i = 1; i < ns.length; i++) {
            if (nums[i] > ns[i - 1]) {
                ns[i] = nums[i];
            } else {
                ns[i] = ns[i - 1];
            }
        }
        for (int i = nums.length - 2; i > -1; i--) {
            if (nums[i] < ms[i + 1]) {
                ms[i] = nums[i];
            } else {
                ms[i] = ms[i + 1];
            }
        }
        int i = 1;
        for (; i < ms.length; i++) {
            if (ms[i] >= ns[i - 1]) {
                break;
            }
        }
        return i;
    }
}
// @lc code=end

File: code/915.分割数组.js
/*
 * @lc app=leetcode.cn id=915 lang=javascript
 *
 * [915] 分割数组
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var partitionDisjoint = function(nums) {
    r = 0
    cur_max = nums[0]
    left_max = nums[0]
    
    for (let i = 1; i < nums.length; i++) {
        cur_max = Math.max(nums[i], cur_max)
        if (left_max > nums[i]) {
            r = i
            left_max = cur_max
        }
    }

    return r + 1 
};
// @lc code=end

